File: packages/cli-python/src/spec_init/managed_region.py
"""Managed-region model + file classes (mirrors cli-node/src/managed-region.ts).

upgrade rewrites only CLI-owned regions of a mixed file (CLAUDE.md), never the
user's content around them. File classes decide per-path whether upgrade
replaces, preserves, or region-swaps a file.
"""

from __future__ import annotations

import re

MANAGED_START = "<!-- throughspec:managed:start -->"
MANAGED_END = "<!-- throughspec:managed:end -->"

_REGION_RE = re.compile(
    r"<!-- throughspec:managed:start -->.*?<!-- throughspec:managed:end -->",
    re.DOTALL,
)

# User-owned data: never overwritten by upgrade (written only if missing).
_PRESERVE = frozenset(
    {
        "claude/context.md",
        "claude/srs.md",
        "claude/plan.md",
        "claude/features.md",
        "claude/design-decisions.md",
        "claude/learnings.md",
        "design/design.md",
        "CHANGELOG.md",
        "README.md",
        "spec.config.js",
    }
)
# Mixed files: CLI instructions inside markers, user data outside.
_MANAGED = frozenset({"CLAUDE.md"})


def classify(rel: str) -> str:
    """Classify a payload-relative path: 'preserve' | 'managed' | 'replace'."""
    if rel in _PRESERVE:
        return "preserve"
    if rel in _MANAGED:
        return "managed"
    return "replace"


def count_regions(text: str) -> int:
    """Count the managed regions in a file."""
    return len(_REGION_RE.findall(text))


def swap_managed_regions(ours: str, theirs: str) -> str | None:
    """Return `ours` with each managed region replaced by `theirs`' region at the
    same position. None when the region counts differ - the caller then preserves
    `ours` rather than risk losing user content."""
    ours_regions = _REGION_RE.findall(ours)
    theirs_regions = _REGION_RE.findall(theirs)
    if not ours_regions or len(ours_regions) != len(theirs_regions):
        return None
    it = iter(theirs_regions)
    return _REGION_RE.sub(lambda _m: next(it), ours)

File: packages/cli-node/src/managed-region.ts
// Managed-region model: upgrade rewrites only CLI-owned regions of a mixed
// file (CLAUDE.md), never the user's content around them. File classes decide
// per-path whether upgrade replaces, preserves, or region-swaps a file.

export const MANAGED_START = '<!-- throughspec:managed:start -->';
export const MANAGED_END = '<!-- throughspec:managed:end -->';

const REGION_RE = /<!-- throughspec:managed:start -->[\s\S]*?<!-- throughspec:managed:end -->/g;

export type FileClass = 'preserve' | 'managed' | 'replace';

// User-owned data: never overwritten by upgrade (written only if missing).
const PRESERVE = new Set<string>([
  'claude/context.md',
  'claude/srs.md',
  'claude/plan.md',
  'claude/features.md',
  'claude/design-decisions.md',
  'claude/learnings.md',
  'design/design.md',
  'CHANGELOG.md',
  'README.md',
  'spec.config.js',
]);

// Mixed files: CLI instructions inside markers, user data outside.
const MANAGED = new Set<string>(['CLAUDE.md']);

/** Classify a payload-relative path for upgrade handling. */
export function classify(rel: string): FileClass {
  if (PRESERVE.has(rel)) return 'preserve';
  if (MANAGED.has(rel)) return 'managed';
  return 'replace';
}

/** Count the managed regions in a file. */
export function countRegions(text: string): number {
  return text.match(REGION_RE)?.length ?? 0;
}

/**
 * Return `ours` with each managed region replaced by the corresponding region
 * from `theirs` (matched by position). Returns null when the two files do not
 * have the same number of managed regions - the caller then preserves `ours`
 * rather than risk losing user content.
 */
export function swapManagedRegions(ours: string, theirs: string): string | null {
  const oursRegions = ours.match(REGION_RE) ?? [];
  const theirsRegions = theirs.match(REGION_RE) ?? [];
  if (oursRegions.length === 0 || oursRegions.length !== theirsRegions.length) return null;
  let i = 0;
  return ours.replace(REGION_RE, () => theirsRegions[i++] as string);
}
